### PR TITLE
Rework for full parameter support

### PR DESCRIPTION
**Pull request by cruelsmith**
#### Pull Request (PR) description
Adding multi instance support by extending the options hash of the puppet parser function to give full parameter support for all needed internal net/ldap settings.

Adds also support for `start_tls` ldap connections. Use `$options = { tls => 'start_tls' }` in that case.

The ldap configuration via the `[master]` section from `puppet.conf` like defined in the README.md is still supported and the new parameter defaults to these options when not provided.

*Example usage:*
```
$filter = '(objectClass=posixAccount)'
$attributes = [
  'dn',
  'uid',
]
$options = {
  server   => 'ldap.example.com'
  port     => 636
  tls      => 'simple_tls'
  cafile   => '/etc/pki/tls/certs.pem'
  base     => 'dc=example,dc=com'
  scope    => 'sub'
  user     => 'cn=puppet,ou=people,dc=example,dc=com'
  password => '<Secure Password>',
}

$ldap_entries = ldapquery($filter, $attributes, $options)
```

#### This Pull Request (PR) fixes the following issues
None

### CODE REVIEW

**Review by alexjfisher — CHANGES_REQUESTED** [review]
I don't think we should make changes before converting the function to a modern API one.
But I 100% agree the options handling needs to be refactored.

Using the modern function API allows us to have more than one dispatch.  This is a convenient way to add functionality without creating any breaking change.  I've made a start in https://github.com/voxpupuli/puppet-ldapquery/pull/45

**Comment by cruelsmith** [comment]
Agree about the refactoring to the new api. So will start to apply my changes based on #45.
The breaking change of the PuppetX class is than gone since is will be removed. The query options Struct will then just be extended.

File: lib/puppet/functions/ldapquery/query.rb [hunk 1]
```diff
@@ -11,9 +11,16 @@
   local_types do
     type <<~TYPE_ALIAS
       Options = Struct[
-        Optional[base]   => String[1],
-        Optional[scope]  => Enum['sub','base','single'],
-        Optional[server] => String[1],
+        Optional[base]           => String[1],
```
Comment (traylenator) [inline lib/puppet/functions/ldapquery/query.rb:14]: Have been looking at this and also #47  which differ in that `base` is it's own parameter.

This looks good other than of course docs and spec but 

There is a case to make scope and base their own parameters so the options becomes only the connection parameters.

**Comment by cruelsmith** [comment]
Closed in favor for #47

**Pull request closed without merging**